community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/CountsRebuilder.java

```diff
@@ -0,0 +1,170 @@
+import org.neo4j.kernel.impl.storemigration.monitoring.MigrationProgressMonitor;
+import org.neo4j.kernel.lifecycle.Lifespan;
+import org.neo4j.logging.NullLogProvider;
+
+import static org.neo4j.kernel.impl.store.MetaDataStore.DEFAULT_NAME;
+import static org.neo4j.kernel.impl.store.format.RecordFormatSelector.selectForVersion;
+import static org.neo4j.kernel.impl.storemigration.FileOperation.DELETE;
+import static org.neo4j.kernel.impl.storemigration.FileOperation.MOVE;
+
+/**
+ * Rebuilds the count store during migration.
+ * <p>
+ * Since the database may or may not reside in the upgrade directory, depending on whether the new format has
+ * different capabilities or not, we rebuild the count store in the store directory if we fail to open the store in
+ * the upgrade directory
+ * <p>
+ * Just one out of many potential participants in a {@link StoreUpgrader migration}.
+ *
+ * @see StoreUpgrader
+ */
+public class CountsRebuilder extends AbstractStoreMigrationParticipant
+{
+
+    public static final Iterable<StoreFile> COUNTS_STORE_FILES = Iterables
+            .iterable( StoreFile.COUNTS_STORE_LEFT, StoreFile.COUNTS_STORE_RIGHT );
+    private final Config config;
+    private final LogService logService;
+    private final FileSystemAbstraction fileSystem;
+    private final PageCache pageCache;
+
+    public CountsRebuilder( FileSystemAbstraction fileSystem, PageCache pageCache, Config config, LogService logService )
+    {
+        super( "Count rebuilding" );
+        this.fileSystem = fileSystem;
+        this.pageCache = pageCache;
+        this.config = config;
+        this.logService = logService;
+    }
+
+    @Override
+    public void migrate( File storeDir, File migrationDir, MigrationProgressMonitor.Section progressMonitor,
+            String versionToMigrateFrom, String versionToMigrateTo ) throws IOException
+    {
+        if ( countStoreRebuildRequired( versionToMigrateFrom ) )
+        {
+            // create counters from scratch
+            StoreFile.fileOperation( DELETE, fileSystem, migrationDir, migrationDir, COUNTS_STORE_FILES, true, null,
+                    StoreFileType.STORE );
+            File neoStore = new File( storeDir, DEFAULT_NAME );
+            long lastTxId = MetaDataStore.getRecord( pageCache, neoStore, Position.LAST_TRANSACTION_ID );
+            try
+            {
+                rebuildCountsFromScratch( migrationDir, lastTxId, progressMonitor, versionToMigrateTo, pageCache );
+            }
+            catch ( StoreFailureException e )
+            {
+                //This means that we did not perform a full migration, as the formats had the same capabilities. Thus
+                // we should use the store directory for information when rebuilding the count store.
+                rebuildCountsFromScratch( storeDir, lastTxId, progressMonitor, versionToMigrateFrom, pageCache );
+            }
+        }
+        progressMonitor.completed();
```

Comment:
This is fixed, I messed up a rename in a rebase though, please look at `CountsMigrator`